src/tasks.ts

```diff
@@ -0,0 +1,329 @@
+import { PubkeyHex } from '@lodestar/api/keymanager';
+import { Epoch, RootHex, Slot } from '@lodestar/types';
+import { SignedBeaconBlock } from '@lodestar/types/allForks';
+import { BlockTag } from 'ethers';
+import * as R from 'ramda';
+
+import { readArtifact, saveArtifact } from './artifacts.js';
+import * as Cache from './cache.js';
+import { catCID, uploadTree } from './ipfs.js';
+import { buildTree, loadTree } from './merkle.js';
+import { shared as Shared } from './shared.js';
+import { ValidatorInfo } from './types.js';
+import { debug, toHex } from './utils.js';
+
+export async function main(): Promise<void> {
+    // TODO: place somewhere else?
+    Cache.destroy();
+
+    const clBlock = await getLastFinalizedBeaconBlock();
+    const blockTag = getBlockTag(clBlock);
+
+    if (!(await isReportable(blockTag))) {
+        debug('Report is not allowed');
+        return;
+    }
+
+    const targetEpoch = await Shared.ORACLE.nextReportEpoch({
+        blockTag,
+    });
+    debug('Target epoch is', targetEpoch);
+
+    if (targetEpoch > slotToEpoch(clBlock.message.slot)) {
+        debug('Target epoch is not reached yet');
+        return;
+    }
+
+    // TODO: query report frame instead
+    const sourceEpoch = await Shared.ORACLE.lastConsolidatedEpoch({
+        blockTag,
+    });
+    debug('Source epoch is', sourceEpoch);
+
+    const artifact = readArtifact(Number(sourceEpoch), Number(targetEpoch));
+    if (!R.isNil(artifact)) {
+        debug('Report tree artifact already exists');
+        return;
+    }
+
+    const prevTreeCID = await Shared.ORACLE.treeCid({
+        blockTag,
+    });
+
+    if (R.isEmpty(prevTreeCID)) {
+        debug('No previous report tree CID found');
+    } else {
+        debug('Reading the previous tree from IPFS by CID', prevTreeCID);
+        const prevTree = loadTree(await catCID(prevTreeCID));
+
+        debug('Check the tree root');
+        const prevRoot = await Shared.ORACLE.reportRoot({
+            blockTag,
+        });
+
+        if (prevTree.root !== prevRoot) {
+            throw new Error('Tree root mismatch');
+        }
+
+        debug('Fill in the rewards cache');
+        for (const [_, [no, reward]] of prevTree.entries()) {
+            Cache.rewards.set(no, reward);
+        }
+    }
+
+    debug('Loading validators');
+    await loadValidators(Number(targetEpoch));
+
+    debug('Fetching duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await lookupCommittees(epoch);
+    }
+
+    debug('Check duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await checkEpochSlots(epoch);
+    }
+
+    // TODO: read the threshold from somewhere
+    Cache.excludeFromStats(belowThreshold(0.9));
```

Comment:
should we store or at least log excluded somewhere?

Reply:
Added logging for debugging purpose.